src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -336,6 +337,12 @@ private static bool AnyArgument(ImmutableArray<IArgumentOperation> arguments, Fu
             // Checks if the parameters of the two passed methods only differ in a ct.
             static bool HasSameParametersPlusCancellationToken(ITypeSymbol cancellationTokenType, IMethodSymbol originalMethod, IMethodSymbol methodToCompare)
             {
+                // Overload is not valid if it has a different return type
+                if (!originalMethod.ReturnType.Equals(methodToCompare.ReturnType))
```

Comment:
I think this is too specific a constraint. The common case will be that the method result is directly awaited,  in which case any awaitable return value is fine. That's probably harder to validate,  but when directly awaited,  we should at least allow for a ValueTask- returning overload where the original is Task- returning.  This is not uncommon,  with a cancellation token being added to a new overload that then also returns value task instead.

Reply:
I think you could amend this to say that if a type is convertible to the resultant type it should be allowed

So we would not warn for something like:

```csharp
static void M1(string s, CancellationToken cancellationToken)
{
    var result = M2(s); // CA2016
}

static long M2(string s) { throw new NotImplementedException(); }

static int M2(string s, CancellationToken cancellationToken) { throw new NotImplementedException(); }
```

but not here:
```csharp
static void M1(string s, CancellationToken cancellationToken)
{
    var result = M2(s); // CA2016
}

static (int value1, int value2) M2(string s) { throw new NotImplementedException(); }

static int M2(string s, CancellationToken cancellationToken) { throw new NotImplementedException(); }
```

I am mostly concerned about cases where the return types are wildly divergent. In most of these cases it is because we are calling a helper type _within_ a method or the methods are simply poorly named and are not semantically equivalent.

Reply:
> I am mostly concerned about cases where the return types are wildly divergent.

Yup, the concern is valid.  Just wanted to avoid losing out on the cases that aren't problematic and that we know do arise. Thanks!

Reply:
I've changed the rules to now consider it a valid overload if:
1. return types are implicitly convertible to each other
2. In the case of generic Task-Like types consider them compatible if their type arguments are implicitly convertable to each other